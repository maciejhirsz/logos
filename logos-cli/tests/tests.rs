use insta::assert_snapshot;
use std::fs::read_to_string;

use assert_cmd::Command;
use assert_fs::{fixture::FileWriteStr, NamedTempFile};

const INPUT_FILE: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/tests/data/input.rs");

fn gen_to_file(format: bool) -> NamedTempFile {
    let tempfile = NamedTempFile::new("output.gen.rs").unwrap();

    let mut cmd = Command::cargo_bin("logos-cli").unwrap();
    cmd.arg(INPUT_FILE).arg("--output").arg(tempfile.path());
    if format {
        cmd.arg("--format");
    }

    cmd.assert().success();

    tempfile
}

fn get_features_label() -> &'static str {
    if cfg!(feature = "state_machine_codegen") {
        "state_machine"
    } else {
        "tailcall"
    }
}

#[test]
fn test_codegen() {
    let tempfile = gen_to_file(false);

    let output = read_to_string(tempfile).expect("Unable to read output file");

    assert_snapshot!(format!("{}-nofmt", get_features_label()), output);
}

#[test]
fn test_codegen_check() {
    let tempfile = gen_to_file(false);

    Command::cargo_bin("logos-cli")
        .unwrap()
        .arg(INPUT_FILE)
        .arg("--check")
        .arg("--output")
        .arg(tempfile.path())
        .assert()
        .success();
}

#[test]
fn test_codegen_check_format() {
    let tempfile = gen_to_file(true);

    Command::cargo_bin("logos-cli")
        .unwrap()
        .arg(INPUT_FILE)
        .arg("--format")
        .arg("--check")
        .arg("--output")
        .arg(tempfile.path())
        .assert()
        .success();
}

#[test]
fn test_codegen_fail_check() {
    let tempfile = NamedTempFile::new("output.gen.rs").unwrap();

    tempfile.write_str("some random data").unwrap();

    Command::cargo_bin("logos-cli")
        .unwrap()
        .arg(INPUT_FILE)
        .arg("--check")
        .arg("--output")
        .arg(tempfile.path())
        .assert()
        .failure();
}

#[test]
fn test_codegen_format() {
    let tempfile = gen_to_file(true);

    let output = read_to_string(tempfile).expect("Unable to read output file");

    assert_snapshot!(format!("{}-fmt", get_features_label()), output);
}
