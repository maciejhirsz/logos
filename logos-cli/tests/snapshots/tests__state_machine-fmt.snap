---
source: logos-cli/tests/tests.rs
expression: output
---
#[derive(Debug, Clone, Copy, PartialEq)]
enum Token {
    Letter,
}
impl<'s> ::logos::Logos<'s> for Token {
    type Error = ();
    type Extras = ();
    type Source = str;
    fn lex(
        lex: &mut ::logos::Lexer<'s, Self>,
    ) -> std::option::Option<std::result::Result<Self, Self::Error>> {
        use logos::internal::{
            CallbackResult, CallbackRetVal, LexerInternal, SkipResult, SkipRetVal,
        };
        use logos::Logos;
        use std::option::Option as _Option;
        use std::result::Result as _Result;
        type _Lexer<'s> = ::logos::Lexer<'s, Token>;
        macro_rules! _fast_loop {
            ($ lex : ident , $ test : ident , $ offset : ident) => {
                'fast_loop: {
                    while let Some(arr) = $lex.read::<&[u8; 8usize]>($offset) {
                        if $test(arr[0usize]) {
                            $offset += 0usize;
                            break 'fast_loop;
                        }
                        if $test(arr[1usize]) {
                            $offset += 1usize;
                            break 'fast_loop;
                        }
                        if $test(arr[2usize]) {
                            $offset += 2usize;
                            break 'fast_loop;
                        }
                        if $test(arr[3usize]) {
                            $offset += 3usize;
                            break 'fast_loop;
                        }
                        if $test(arr[4usize]) {
                            $offset += 4usize;
                            break 'fast_loop;
                        }
                        if $test(arr[5usize]) {
                            $offset += 5usize;
                            break 'fast_loop;
                        }
                        if $test(arr[6usize]) {
                            $offset += 6usize;
                            break 'fast_loop;
                        }
                        if $test(arr[7usize]) {
                            $offset += 7usize;
                            break 'fast_loop;
                        }
                        $offset += 8usize;
                    }
                    while let Some(byte) = $lex.read::<u8>($offset) {
                        if $test(byte) {
                            break 'fast_loop;
                        }
                        $offset += 1;
                    }
                }
            };
        }
        #[inline]
        fn _make_error<'s>(lex: &mut _Lexer<'s>) -> <Token as Logos<'s>>::Error {
            <Token as Logos<'s>>::Error::default()
        }
        #[derive(Clone, Copy)]
        enum LogosState {
            state0,
            state1,
            state2,
            state3,
            state4,
        }
        let mut state = LogosState::state0;
        let mut offset = lex.offset();
        loop {
            match state {
                LogosState::state0 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (byte == b'a') {
                            offset += 1;
                            state = LogosState::state1;
                            continue;
                        }
                    } else {
                        if lex.offset() == offset {
                            return None;
                        }
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
                LogosState::state1 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (byte == b'-') {
                            offset += 1;
                            state = LogosState::state2;
                            continue;
                        }
                    } else {
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
                LogosState::state2 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                        if (byte == b'z') {
                            offset += 1;
                            state = LogosState::state3;
                            continue;
                        }
                    } else {
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
                LogosState::state3 => {
                    lex.end(offset);
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                    } else {
                    }
                    return Some(Ok(Token::Letter));
                }
                LogosState::state4 => {
                    let other = lex.read::<u8>(offset);
                    if let Some(byte) = other {
                    } else {
                    }
                    return Some(Ok(Token::Letter));
                }
            }
        }
    }
}
